hamilton/lifecycle/default.py

```diff
@@ -348,9 +348,11 @@ def __init__(
             key=CacheAdapter.nodes_history_key, default=dict()
         )
         self.used_nodes_hash: Dict[str, str] = dict()
+        self.cache.close()
 
     def run_before_graph_execution(self, *, graph: HamiltonGraph, **kwargs):
         """Set `cache_vars` to all nodes if received None during `__init__`"""
+        self.cache = shelve.open(self.cache_path)
```

Comment:
Something feels wrong about this. Aren't we just opening it again? Feels like it should:
1. Open during `init`
2. Run during execution
3. Close at the end

Reply:
well, that was the previous behavior... I want to have `.close()` *somewhere* for safety, but putting it in `run_after_graph_execution` means you can only run it once

Reply:
I think that's what Thierry just did here?

Reply:
Ahh yes, my bad